### PR TITLE
Translate to IN + subquery when possible instead of EXISTS

### LINKED ISSUE #30955: Translate Contains to IN with subquery instead of EXISTS where relevant
Although our InExpression supports having a subquery (and various visitors handle that), we never actually generate this  - instead we always generate an EXISTS subquery with a predicate (based on SQL Server baselines). Translating to InExpression is likely better - it doesn't require a correlated subquery and expresses the intent more succintly, without needing a predicate. For example, instead of this:

```sql
SELECT ...
FROM [PrimitiveCollectionsEntity] AS [p]
WHERE EXISTS (
    SELECT 1
    FROM (
        SELECT TOP(2) CAST([i].[value] AS int) AS [c], CAST([i].[key] AS int) AS [c0]
        FROM OPENJSON([p].[Ints]) AS [i]
        ORDER BY CAST([i].[key] AS int)
    ) AS [t]
    WHERE [t].[c] = 11)
```

We can translate to this:

```sql
SELECT ...
FROM [PrimitiveCollectionsEntity] AS [p]
WHERE 11 IN (
    SELECT TOP(2) CAST([i].[value] AS int)
    FROM OPENJSON([p].[Ints]) AS [i]
    ORDER BY CAST([i].[key] AS int)
)
```    

However, note that our null semantics around InExpression with subquery currently seems broken (fortunately it's dead code): IN returns null when the item is null (or when the values contains null and a match isn't found). We'd need to make that logic actually work.

### PR DESCRIPTION
**Pull request by roji**
~This builds on top of #30956 (ignore first commit)~ to translate to IN+subquery where possible (see #30955).

Closes #30955

### CODE REVIEW

File: src/EFCore.Relational/Query/RelationalSqlTranslatingExpressionVisitor.cs [hunk 1]
```diff
@@ -118,33 +118,38 @@ protected virtual void AddTranslationErrorDetails(string details)
     ///     Translates an expression to an equivalent SQL representation.
     /// </summary>
     /// <param name="expression">An expression to translate.</param>
+    /// <param name="applyDefaultTypeMapping">
+    ///     Whether to apply the default type mapping on the top-most element if it has none. Defaults to <see langword="false" />.
+    /// </param>
     /// <returns>A SQL translation of the given expression.</returns>
-    public virtual SqlExpression? Translate(Expression expression)
+    public virtual SqlExpression? Translate(Expression expression, bool applyDefaultTypeMapping = true)
```
Comment (roji) [inline src/EFCore.Relational/Query/RelationalSqlTranslatingExpressionVisitor.cs:125]: Note this change: it's needed to allow translating e.g. the item of InExpression and then to infer its type mapping from the subquery projection (otherwise the item gets the default type mapping based on its CLR type). We'll have more uses for this later.

File: src/EFCore.Relational/Query/SqlNullabilityProcessor.cs [hunk 2]
```diff
@@ -651,6 +651,9 @@ protected virtual SqlExpression VisitIn(InExpression inExpression, bool allowOpt
     {
         var item = Visit(inExpression.Item, out var itemNullable);
 
+        // If the InExpression is negated, it's as if we have an enclosing NOT, which prohibits optimized expansion.
+        allowOptimizedExpansion &= !inExpression.IsNegated;
```
Comment (roji) [inline src/EFCore.Relational/Query/SqlNullabilityProcessor.cs:655]: @maumar note the new, full handling for InExpression+subquery, your null semantics foo would be greatly appreciated :)

**Comment by roji** [comment]
Given the pretty massive perf impact this has (see https://github.com/dotnet/efcore/issues/30955#issuecomment-1566951015), converting to a draft to consider using IN in more scenarios instead of rewriting to EXISTS (but this can still be reviewed already).

File: test/EFCore.SqlServer.FunctionalTests/Query/ComplexNavigationsQuerySqlServerTest.cs [hunk 3]
```diff
@@ -2013,10 +2013,11 @@ public override async Task Contains_with_subquery_optional_navigation_scalar_dis
 SELECT [l].[Id], [l].[Date], [l].[Name], [l].[OneToMany_Optional_Self_Inverse1Id], [l].[OneToMany_Required_Self_Inverse1Id], [l].[OneToOne_Optional_Self1Id]
 FROM [LevelOne] AS [l]
 LEFT JOIN [LevelTwo] AS [l0] ON [l].[Id] = [l0].[Level1_Optional_Id]
-WHERE EXISTS (
-    SELECT DISTINCT 1
+WHERE 1 IN (
+    SELECT DISTINCT CAST(LEN([l1].[Name]) AS int)
```
Comment (maumar) [inline test/EFCore.SqlServer.FunctionalTests/Query/ComplexNavigationsQuerySqlServerTest.cs:2017]: we could nuke distinct here (could be done as separate optimization)

File: test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs [hunk 4]
```diff
@@ -3973,6 +3975,8 @@ public virtual async Task Nested_contains_with_enum()
             var keys = new List<Guid> { Guid.Parse("0a47bcb7-a1cb-4345-8944-c58f82d6aac7"), key };
             var todoTypes = new List<MyContext12732.TodoType> { MyContext12732.TodoType.foo0 };
 
+            // Note that in thie query, the outer Contains really has no type mapping, neither for its source (collection parameter), nor
```
Comment (maumar) [inline test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs:3978]: typo

File: test/EFCore.SqlServer.FunctionalTests/Query/ComplexNavigationsQuerySqlServerTest.cs (hunk 3 above)
Comment (roji) [reply test/EFCore.SqlServer.FunctionalTests/Query/ComplexNavigationsQuerySqlServerTest.cs:2017]: Yep, you're right - for all forms of IN, EXISTS, set operations... Opened #31016.

File: test/EFCore.SqlServer.FunctionalTests/Query/ComplexNavigationsSharedTypeQuerySqlServerTest.cs [hunk 5]
```diff
@@ -571,7 +571,11 @@ public override async Task Let_let_contains_from_outer_let(bool async)
     WHERE [t].[OneToOne_Required_PK_Date] IS NOT NULL AND [t].[Level1_Required_Id] IS NOT NULL AND [t].[OneToMany_Required_Inverse2Id] IS NOT NULL AND [t0].[Level2_Required_Id] IS NOT NULL AND [t0].[OneToMany_Required_Inverse3Id] IS NOT NULL AND EXISTS (
         SELECT 1
         FROM [Level1] AS [l3]
-        WHERE [l3].[OneToOne_Required_PK_Date] IS NOT NULL AND [l3].[Level1_Required_Id] IS NOT NULL AND [l3].[OneToMany_Required_Inverse2Id] IS NOT NULL AND [l].[Id] = [l3].[OneToMany_Required_Inverse2Id] AND [l3].[Id] = [t0].[Level2_Required_Id])
+        WHERE [l3].[OneToOne_Required_PK_Date] IS NOT NULL AND [l3].[Level1_Required_Id] IS NOT NULL AND [l3].[OneToMany_Required_Inverse2Id] IS NOT NULL AND [l].[Id] = [l3].[OneToMany_Required_Inverse2Id] AND (CASE
```
Comment (maumar) [inline test/EFCore.SqlServer.FunctionalTests/Query/ComplexNavigationsSharedTypeQuerySqlServerTest.cs:574]: this one is a bit fishy. I assume it fixes some small unknown bug, where we didn't do shared type expansion correctly deep inside a subquery etc, but perhaps we are over eager here

File: src/EFCore.Relational/Query/RelationalSqlTranslatingExpressionVisitor.cs (hunk 1 above)
Comment (maumar) [reply src/EFCore.Relational/Query/RelationalSqlTranslatingExpressionVisitor.cs:125]: please add test(s) that excerises this functionality, i.e. where collection parameter infers some funky mapping from the subquery (perhaps some sort of custom converter?

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 6]
```diff
@@ -1771,10 +1792,13 @@ static Expression GetEntitySource(IModel model, Expression propertyAccessExpress
     ///     Translates the given expression into equivalent SQL representation.
     /// </summary>
     /// <param name="expression">An expression to translate.</param>
+    /// <param name="applyDefaultTypeMapping">
+    ///     Whether to apply the default type mapping on the top-most element if it has none. Defaults to <see langword="false" />.
```
Comment (maumar) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:1796]: defaults to true in code though

File: src/EFCore.Relational/Query/RelationalSqlTranslatingExpressionVisitor.cs [hunk 7]
```diff
@@ -118,33 +118,38 @@ protected virtual void AddTranslationErrorDetails(string details)
     ///     Translates an expression to an equivalent SQL representation.
     /// </summary>
     /// <param name="expression">An expression to translate.</param>
+    /// <param name="applyDefaultTypeMapping">
+    ///     Whether to apply the default type mapping on the top-most element if it has none. Defaults to <see langword="false" />.
```
Comment (maumar) [inline src/EFCore.Relational/Query/RelationalSqlTranslatingExpressionVisitor.cs:122]: same discrepancy here (true in code, false in comment)

File: src/EFCore.Relational/Query/SqlNullabilityProcessor.cs (hunk 2 above)
Comment (maumar) [reply src/EFCore.Relational/Query/SqlNullabilityProcessor.cs:655]: null semantics part looks great!

File: test/EFCore.SqlServer.FunctionalTests/Query/ComplexNavigationsSharedTypeQuerySqlServerTest.cs (hunk 5 above)
Comment (roji) [reply test/EFCore.SqlServer.FunctionalTests/Query/ComplexNavigationsSharedTypeQuerySqlServerTest.cs:574]: OK, I figured this out.

Here's the relevant tree fragment that arrives to SqlNullabilityProcessor before this PR (with ExistsExpression):

```sql
EXISTS (
    SELECT 1
    FROM Level1 AS l3
    WHERE (((NotEqual(l3.OneToOne_Required_PK_Date) && NotEqual(l3.Level1_Required_Id)) && NotEqual(l3.OneToMany_Required_Inverse2Id)) && ((l.Id != NULL) && (l.Id == l3.OneToMany_Required_Inverse2Id)))
        && (CASE WHEN ((l3.OneToOne_Required_PK_Date != NULL) && (l3.Level1_Required_Id != NULL)) && (l3.OneToMany_Required_Inverse2Id != NULL) THEN l3.Id END == t0.Level2_Required_Id))
```

And here's the same piece of the tree after this PR (with InExpression):

```sql
t0.Level2_Required_Id IN (
    SELECT CASE
        WHEN ((l3.OneToOne_Required_PK_Date != NULL) && (l3.Level1_Required_Id != NULL)) && (l3.OneToMany_Required_Inverse2Id != NULL) THEN l3.Id
    END
    FROM Level1 AS l3
    WHERE ((NotEqual(l3.OneToOne_Required_PK_Date) && NotEqual(l3.Level1_Required_Id)) && NotEqual(l3.OneToMany_Required_Inverse2Id)) && ((l.Id != NULL) && (l.Id == l3.OneToMany_Required_Inverse2Id)))
```

Note that the same CASE is present in both - but with IN it's in the projection, whereas with EXISTS it's in the predicate. With EXISTS, since it's the last thing in the predicate, the previous checks insert the 3 columns (e.g. OneToOne_Required_PK_Date) into _nonNullableColumns, so when we process the CASE everything is elided. But with IN, since it's not part of the predicate, _nonNullableColumns is empty and the full CASE gets generated. Since both the item and the subquery of the IN are nullable, we transform it to EXISTS, moving the projection (with the CASE) into the predicate, so you get a complicated EXISTS.

One way to fix this is to visit the new ExistsExpression after we create it from the InExpression - this simplifies the SQL since the predicate now contains the null checks before the CASE (just like before); but it means double visitation.

The root problem here is that _nonNullableColumns isn't used more extensively; when processing SelectExpression, _nonNullableColumns coming out of the predicate should simply affect the predicate (and also other parts of the SelectExpression, like ORDER BY). Have we considered doing something like this?

I think it's fine to leave this as-is - there's no bug here, just slightly less efficient SQL for this specific edge case (and far more efficient SQL for lots of other cases). We can tackle using _nonNullableColumns more extensively - what do you think?

Comment (maumar) [reply test/EFCore.SqlServer.FunctionalTests/Query/ComplexNavigationsSharedTypeQuerySqlServerTest.cs:574]: I agree to leave as is. Filed #31022 for general purpose improvement to _nonNullableColumns

File: src/EFCore.Relational/Query/RelationalSqlTranslatingExpressionVisitor.cs (hunk 1 above)
Comment (roji) [reply src/EFCore.Relational/Query/RelationalSqlTranslatingExpressionVisitor.cs:125]: Added CustomConvertersTestBase.Infer_type_mapping_from_in_subquery_to_item

**Review by maumar — APPROVED** [review]
